plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java

```diff
@@ -67,47 +69,34 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
-                        for (Map<String, Object> entry : entries) {
-                            // Create a new resource for the entry
-                            final ImmutableCollectionSetResource.Builder<GenericTypeResource> resourceBuilder = ImmutableCollectionSetResource.newBuilder(GenericTypeResource.class)
-                                    .setResource(ImmutableGenericTypeResource.newBuilder()
-                                            .setNodeResource(nodeResource)
-                                            .setType(resourceType.getName())
-                                            .setInstance(resourceTypeGenerator.getIndex(rule, entry))
-                                            .build());
-
-                            // Process the items in the rule
-                            boolean didAddAttribute = false;
-                            for (Item item : rule.getItemPrototypes()) {
-                                final String effectiveKey = ZabbixMacroSupport.evaluateMacro(item.getKey(), entry);
-                                String value = client.retrieveData(effectiveKey).get();
-                                addValueToMapper(rule, item, value, resourceBuilder);
-                                didAddAttribute = true;
-                            }
-
-                            // Only add the resource if we found 1+ items
-                            if (didAddAttribute) {
-                                collectionSetBuilder.addCollectionSetResource(resourceBuilder.build());
-                            }
+                processItems(client, null, template.getItems(), nodeResourceBuilder).thenRunAsync(() -> {
+                    final ZabbixResourceTypeGenerator resourceTypeGenerator = new ZabbixResourceTypeGenerator();
+                    List<CompletableFuture<Void>> ruleFutureList = new ArrayList<>();
+                    for (DiscoveryRule rule : template.getDiscoveryRules()) {
+                        try {
+                            ruleFutureList.add(client.discoverData(rule.getKey())
+                                    .thenAccept(list -> {
+                                        try {
+                                            processEntries(client, rule, list, collectionSetBuilder, nodeResource, resourceTypeGenerator);
+                                        } catch (IOException e) {
+                                            e.printStackTrace();
+                                        }
+                                    }));
+                        } catch (ZabbixNotSupportedException | IOException e) {
+                            // pass
                         }
-                    } catch (ZabbixNotSupportedException e) {
-                        // pass
                     }
-                }
+                    try {
+                        CompletableFuture.allOf(ruleFutureList.toArray(new CompletableFuture[ruleFutureList.size()])).get();
```

Comment:
We shouldn't block here - when these futures are complete, we should be able to complete the future that is returned by the function.